client/app/src/main/java/com/sciaps/apidemo/SciApsClient.java

```diff
@@ -2,76 +2,296 @@
     public static final int ANALYZER_PORT = 8080;
 
     private final Gson mGson;
-    private static HttpPost mAcquirePostRequest;
-    private final HttpClient mHttpClient;
-    private String mIPaddress;
+    private final CloseableHttpClient mHttpClient;
     private InetSocketAddress mInetSocketAddress;
 
     public SciApsClient(final String ipaddress) {
-        mGson = new GsonBuilder().create();
+        mInetSocketAddress = new InetSocketAddress(ipaddress, ANALYZER_PORT);
+        mGson = new GsonBuilder()
+                .serializeSpecialFloatingPointValues()
+                .serializeNulls()
+                .create();
+
+        RequestConfig config = RequestConfig.custom()
+                .setConnectTimeout(DEFAULT_CONNECTION_TIMEOUT)
+                .setSocketTimeout(DEFAULT_SOCKET_TIMEOUT)
+                .build();
 
         mHttpClient = HttpClientBuilder.create()
-                .setMaxConnTotal(5)
+                .setConnectionManager(new BasicHttpClientConnectionManager())
+                .setDefaultRequestConfig(config)
                 .build();
-        mIPaddress = ipaddress;
-        mInetSocketAddress = new InetSocketAddress(ipaddress, ANALYZER_PORT);
+    }
+
+    private String getUrlForEndpointWithArgs(final String endpoint, final String mode, final String spectra) {
+        String url = String.format("http://%s/api/v2/%s", mInetSocketAddress.getHostString(), endpoint);
+        if (spectra != null) {
+            url += "/" + spectra;
+        }
+        if (mode != null) {
+            url += "?mode=" + mode;
+        }
+        return url;
     }
 
     private String getUrlForEndpoint(final String endpoint) {
-        return String.format("http://%s:%d/api/v2/%s", mInetSocketAddress.getHostString(), mInetSocketAddress.getPort(), endpoint);
+        return getUrlForEndpointWithArgs(endpoint, null, null);
     }
 
-    public InstrumentConfig getInstrumentConfig() {
+    private String getUrlForEndpointWithMode(final String endpoint, final String mode) {
+        return getUrlForEndpointWithArgs(endpoint, mode, null);
+    }
 
-        String url = getUrlForEndpoint("/config");
-        HttpGet request = new HttpGet(url);
+    private String getUrlForEndpointWithModeAndSpecta(final String endpoint, final String mode, boolean allSpectra) {
+        return getUrlForEndpointWithArgs(endpoint, mode, allSpectra ? "all" : "final");
+    }
 
-        RequestConfig requestConfig = RequestConfig.custom()
-                .setConnectTimeout(DEFAULT_CONNECTION_TIMEOUT)
-                .setSocketTimeout(DEFAULT_SOCKET_TIMEOUT)
-                .build();
+    public InstrumentId getInstrumentId() {
+        InstrumentId retval = null;
+        HttpGet request = new HttpGet(getUrlForEndpoint("id"));
+        try (CloseableHttpResponse response = mHttpClient.execute(request)) {
```

Comment:
Hi @ftr-sciaps - This try-catch block is duplicated in pretty much every method. We should refactor it to avoid duplication.

Reply:
yes, and it's needed to ensure the response is closed.  I don't see how to factor this out without introducing generics or something more complex.  This is a simple client meant to show proper usage of the third party api and is intentionally kept simple and straight forward.

Reply:
There is duplicated code in every one of these APIs. It can be made simpler that what it is right now.